internal/allocator/allocator.go

```diff
@@ -245,89 +289,222 @@ func (a *Allocator) Unassign(svc string) {
 	stats.ipv6PoolActive.WithLabelValues(al.pool).Set(float64(len(a.poolIPV6InUse[al.pool])))
 }
 
-// AllocateFromPool assigns an available IP from pool to service.
-func (a *Allocator) AllocateFromPool(svcKey string, svc *v1.Service, serviceIPFamily ipfamily.Family, poolName string, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
-	if alloc := a.allocated[svcKey]; alloc != nil {
-		// Handle the case where the svc has already been assigned an IP but from the wrong family.
-		// This "should-not-happen" since the "serviceIPFamily" is an immutable field in services.
-		allocIPsFamily, err := ipfamily.ForAddressesIPs(alloc.ips)
-		if err != nil {
-			return nil, err
-		}
-		if allocIPsFamily != serviceIPFamily {
-			return nil, fmt.Errorf("IP for wrong family assigned alloc %s service family %s", allocIPsFamily, serviceIPFamily)
-		}
-		if err := a.Assign(svcKey, svc, alloc.ips, ports, sharingKey, backendKey); err != nil {
-			return nil, err
-		}
-		return alloc.ips, nil
-	}
-
-	pool := a.pools.ByName[poolName]
-	if pool == nil {
-		return nil, fmt.Errorf("unknown pool %q", poolName)
-	}
-
-	ips := []net.IP{}
-	ipfamilySel := make(map[ipfamily.Family]bool)
-
-	switch serviceIPFamily {
-	case ipfamily.DualStack:
-		ipfamilySel[ipfamily.IPv4], ipfamilySel[ipfamily.IPv6] = true, true
-	default:
-		ipfamilySel[serviceIPFamily] = true
-	}
-
+// getFreeIPsFromPool determines, with best effort, an ipv4 and an ipv6 available from the provided pool.
+func (a *Allocator) getFreeIPsFromPool(
```

Comment:
not sure I understand how this function returns free ips, and not just some ips in the pool (that can already be taken)?